### PR TITLE
[qwen2 vl] fix packing with all attentions

### LINKED ISSUE #39400: Causal mask is not compatible with Qwen2-VL when using padding-free training
### System Info

transformers @ 4.53.2

### Who can help?

@amyeroberts @zucchini-nlp 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

https://github.com/huggingface/transformers/blob/903944a411c35b8f7b2b51ada77a7c2a80e7fb88/src/transformers/masking_utils.py#L708-L713

Qwen2-VL's position ids has a shape of `[3, batch_batch, seq_len]`, while the current code cannot process the 3D position ids.

https://github.com/hiyouga/EasyR1/issues/413

### Expected behavior

There should be a special case for Qwen2-VL's 3D MRope

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, and let's merge https://github.com/huggingface/transformers/pull/39341/files first. Added tests to make sure it works. If users want to use packing with any of the attentions, it is their responsibility to prepare correct position ids, since Qwen-VL needs positions for all vision grids

Fixes https://github.com/huggingface/transformers/issues/39400, https://github.com/huggingface/transformers/issues/38007

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_vl

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_omni, qwen2_5_vl, qwen2_vl

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_omni, qwen2_5_vl, qwen2_vl

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -350,6 +350,7 @@ def _prepare_flash_attention_from_position_ids(query, key, value, position_ids):
     # This is a limitation of flash attention API, as the function `flash_attn_varlen_func`
     # requires `max_length_q`, `max_length_k` to be passed as `int` and not `torch.Tensor`.
     # https://github.com/Dao-AILab/flash-attention/blob/2dd8078adc1d9b74e315ee99718c0dea0de8eeb6/flash_attn/flash_attn_interface.py#L1423-L1424
+
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py]:
```suggestion
```

File: src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py [hunk 2]
```diff
@@ -844,8 +819,35 @@ def prepare_inputs_for_generation(
...
 
-        # Qwen2-5-VL position_ids are prepareed with rope_deltas in forward
-        model_inputs["position_ids"] = None
+        # Qwen2-5-VL position_ids are prepared with rope_deltas
+        if position_ids is None:
+            # Calculate RoPE index once per generation in the pre-fill stage only.
+            # When compiling, we can't check tensor values thus we check only input length
+            # It is safe to assume that `length!=1` means we're in pre-fill because compiled
+            # models currently cannot do asssisted decoding
+            if cache_position[0] == 0 or self.model.rope_deltas is None:
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py:843]: Small q if this is compile compatible or not!

File: src/transformers/models/qwen2_vl/modeling_qwen2_vl.py [hunk 3]
```diff
@@ -1217,44 +1233,23 @@ def forward(
...
                 self.rope_deltas = rope_deltas
             # then use the prev pre-calculated rope-deltas to get the correct position ids
             else:
                 batch_size, seq_length, _ = inputs_embeds.shape
-                delta = cache_position[0] + self.rope_deltas if cache_position is not None else 0
+                delta = (
+                    (cache_position[0] + self.rope_deltas).to(inputs_embeds.device)
+                    if cache_position is not None
+                    else torch.zeros((batch_size, seq_length), device=inputs_embeds.device)
+                )
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_vl/modeling_qwen2_vl.py]: prefer a if else than this!

**Review by ArthurZucker — COMMENTED** [review]
Looks cleaner! Can you add a small test? very nice otherwise

File: src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py:843]: no it's not, good catch, reverting back! It won't fail for generation, because we will prepare position ids in advance, but if someone does simple forward we need the workaround

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_omni, qwen2_5_vl, qwen2_vl

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_omni, qwen2_5_vl, qwen2_vl

**Review by ArthurZucker — APPROVED** [review]
#39474 kinda related, updating generate to properly prepare the inputs

**Comment by zucchini-nlp** [comment]
Thanks, let's merge and I will fix SDPA/eager packing tests for all other models in subsequent PRs. We never had tests so it's failing for ~10-15 models now 👀  I marked it as slow to not ruin CI

**Comment by hebiao064** [comment]
When can we expect a new release of transformers that includes this fix?

**Comment by zucchini-nlp** [comment]
Release is planned for tomorrow

**Comment by hebiao064** [comment]
> Release is planned for tomorrow

That would be great! VERL + SGLang Community really need it, thanks a lot!
